### ISSUE TITLE
[BUG] vm-dhcp-controller dhcpserver-agent Pod Fails to Start in Air-Gapped Environment Due to ImagePullPolicy 'Always' Setting

### ISSUE THREAD

**Issue by zha0jf** [harvester/harvester#6942]
**Describe the bug**  
In the `prepareAgentPod` function within the file `vm-dhcp-controller/pkg/controller/ippool/common.go`, an init container is added to the `vm-dhcp-agent` pod, using the image `docker.io/library/busybox` with the default `imagePullPolicy` set to `Always`. This configuration causes the vm-dhcp-agent's `dhcpserver-agent` pod to get stuck in the “Init:ImagePullBackOff” status when deploying in an air-gapped environment, even if the `busybox:latest` image was pre-uploaded offline. This prevents the `vm-dhcp-agent` from functioning in air-gapped environments.

**To Reproduce**  
1. Deploy `vm-dhcp-controller` in an air-gapped environment.
2. Even after uploading the `busybox:latest` image offline, the vm-dhcp-agent `dhcpserver-agent` pod remains stuck in the "Init:ImagePullBackOff" status.

**Expected behavior**  
The vm-dhcp-agent `dhcpserver-agent` pod should be able to pull the locally uploaded image and start successfully in an air-gapped environment.

**Support bundle**  

**Environment**  
- Harvester ISO version:  harvester-v1.4.0-rc4-amd64.iso

**Additional context**  
It is recommended to add the configuration `ImagePullPolicy = "IfNotPresent"` at line 96 in the file `vm-dhcp-controller/pkg/controller/ippool/common.go` to ensure that the `vm-dhcp-controller` can function properly in air-gapped environments.

**Comment by starbops** [comment]
Hi @zha0jf, thanks for reporting. I wonder why, even though the missing container image is imported to the node, it still complains about `Init:ImagePullBackOff`. Did you wait enough time to surpass the back-off time? Another simple way to test it is to simply delete the DHCP agent pod entirely.

**Comment by starbops** [comment]
`imagePullPolicy` defaults to `Always` because we specify the `busybox` image without a tag. [As the document explains](https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy), kubelet will check if the digest exists in the cache. This involves communication between kubelet and remote registries. That might be why you constantly get an `Init:ImagePullBackOff` error even though the container image is manually imported.

Your fix in harvester/vm-dhcp-controller#37 can resolve the issue. However, this also reminds us that we should use any container image packed in the ISO image. I'd say that's the root cause of the problem. I'm considering adding iproute2 to the `rancher/harvester-vm-dhcp-agent` image and making the init container use it. We'll take care of that in following PRs. Thank you.

**Comment by zha0jf** [comment]
That will be great. Thank you.

**Comment by starbops** [comment]
Hi @zha0jf, We've published harvester-vm-dhcp-controller 0.3.3. Would you like to try it to see if it solves the problem?

https://github.com/harvester/experimental-addons/blob/main/harvester-vm-dhcp-controller/harvester-vm-dhcp-controller.yaml

Thank you.

**Comment by zha0jf** [comment]
Hi @starbops，I just tested harvester-vm-dhcp-controller 0.3.3 in the harvester-v1.4.0-rc5 air-gapped environment, and the issue has been resolved. Thank you.

**Comment by starbops** [comment]
@zha0jf That's a great news! Thanks again for spotting the issue and sending a patch to us :)

Note: please don't close this issue yet. We have our pipelines and will take care of it.

**Comment by irishgordo** [comment]
@starbops w/ the vm-dhcp-controller is there a list of images, like an image.txt that would need to be pulled into a private docker registry prior?  Alongside, hosting the chart?

**Comment by TachunLin** [comment]
We tried to install the `harvester-vm-dhcp-controller` addon in the air-gapped environment. 

Refer to the [document steps](https://docs.harvesterhci.io/v1.4/advanced/addons/managed-dhcp/#install-and-enable-the-vm-dhcp-controller-add-on) to create the yaml file locally from copy the content. 


It can deploy the `vm-dhcp-controller` addon, but when we try to enable, it stuck in Enabling with error message

  ![Image](https://github.com/user-attachments/assets/0fe3915a-8341-4796-a3bb-b0a3e2785157)
  

  ![Image](https://github.com/user-attachments/assets/9cdc4caf-0983-40f1-a70b-6febb1c1e649)
  
  https://github.com/user-attachments/assets/d5550e15-13b9-46e2-8a43-1a531ea8eccf




After discussion, we might need to prepare the private registry and chart server inside the air-gapped environment to correctly deploy and enable the addon.

**Comment by TachunLin** [comment]
1. We build up an airgapped environment using [vagrant-pxe-airgap-harvester](https://github.com/harvester/ipxe-examples/tree/main/vagrant-pxe-airgap-harvester)

2. And follow the steps to create the private registry on the pxe_server
https://github.com/harvester/tests/issues/347#issue-1241140276

3. Then we pull and push the harvester-vm-dhcp addon related images to the private registry

```
docker pull rancher/harvester-vm-dhcp-controller:v0.3.3
docker pull rancher/harvester-vm-dhcp-webhook:v0.3.3
docker pull rancher/harvester-vm-dhcp-agent:v0.3.3

docker tag rancher/harvester-vm-dhcp-webhook:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
docker push myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3

docker tag rancher/harvester-vm-dhcp-webhook:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
docker push myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3

docker tag rancher/harvester-vm-dhcp-agent:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-agent:v0.3.3
docker push myregistry.local:5000/rancher/harvester-vm-dhcp-agent:v0.3.3

```

4. Clone the https://github.com/harvester/charts 

5. Checkout to the v0.3.3 

```
cd charts/
git checkout tags/harvester-vm-dhcp-controller-0.3.3 -b harvester-vm-dhcp-controller-0.3.3
mkdir serv
helm package . -d serv
helm repo index serv
cd serv/
python3 -m http.server -b 0.0.0.0
```

6. Create the harvester-vm-dhcp-controller addon from yaml

```
apiVersion: harvesterhci.io/v1beta1
kind: Addon
metadata:
  name: harvester-vm-dhcp-controller
  namespace: harvester-system
  labels:
    addon.harvesterhci.io/experimental: "true"
spec:
  enabled: false
  repo: http://192.168.2.254:8000
  version: 0.3.3
  chart: harvester-vm-dhcp-controller
  valuesContent: |
    image:
      repository: myregistry.local:5000/rancher/harvester-vm-dhcp-controller
    webhook:
      image:
        repository: myregistry.local:5000/rancher/harvester-vm-dhcp-webhook
    agent:
      image:
        repository: myregistry.local:5000/rancher/harvester-vm-dhcp-agent
```


7. But  encounter failure while enabling the `vm-dhcp-conroller` addon
    - Pod failure 
      ![Image](https://github.com/user-attachments/assets/b0ebe0fa-1681-4b87-b026-2fb9dedfe5ad)
    
    - pod logs  
      ![Image](https://github.com/user-attachments/assets/95d4b782-c64a-4a97-920b-8ba22ef2323f)
      
    - pod Events
    ![Image](https://github.com/user-attachments/assets/b67dd856-f7a8-4365-8c85-bfc36e9a6f0a)
    
    ```
     Events:                                                                                                                                                    │
    │   Type     Reason          Age                    From               Message                                                                               │
    │   ----     ------          ----                   ----               -------                                                                               │
    │   Normal   Scheduled       53m                    default-scheduler  Successfully assigned harvester-system/harvester-vm-dhcp-controller-694599df49-v6bw6  │
    │ to harvester-node-0                                                                                                                                        │
    │   Normal   AddedInterface  53m                    multus             Add eth0 [10.52.0.74/32] from k8s-pod-network                                         │
    │   Normal   Pulled          51m (x5 over 53m)      kubelet            Container image "myregistry.local:5000/rancher/harvester-vm-dhcp-controller:v0.3.3" a │
    │ lready present on machine                                                                                                                                  │
    │   Normal   Created         51m (x5 over 53m)      kubelet            Created container harvester-vm-dhcp-controller                                        │
    │   Normal   Started         51m (x5 over 53m)      kubelet            Started container harvester-vm-dhcp-controller                                        │
    │   Warning  BackOff         3m18s (x233 over 53m)  kubelet            Back-off restarting failed container harvester-vm-dhcp-controller in pod harvester-vm │
    │ -dhcp-controller-694599df49-v6bw6_harvester-system(7c691cab-a48c-4886-a9ae-8e52919fe3fe)
    ```

**Comment by w13915984028** [comment]
@TachunLin It looks the image was pulled, but container was terminated due to other reasons. Do you have a SB file?

**Comment by TachunLin** [comment]
The issue was found when me and Zespre working together to configure the private registry and chart server.
The possible root cause is the way addon pods handle with the remote URL path string.

**Comment by TachunLin** [comment]
We can use another manner to bypass the failure and correctly install the vm-dhcp-controller addon under `air-gapped` environment.  

  ![Image](https://github.com/user-attachments/assets/01c493d5-5057-4ed6-99c0-d3ecc0dc7389)
  
  ![Image](https://github.com/user-attachments/assets/5fb9a654-d250-41dd-b57b-ef3183ff118f)

Following is the detailed steps for the reference:

1. Provision the air-gapped Harvester with [vagrant-pxe-airgap-harvester ](https://github.com/harvester/ipxe-examples/tree/main/vagrant-pxe-airgap-harvester)

2. Access the pxe_server (ssh vagrant@192.168.2.254) , change to root user, install docker and helm 

3. Refer to the steps to create a private registry with self-signed ca on the pxe_server
    https://github.com/harvester/tests/issues/347

    ```
    mkdir -p certs
    openssl req \
      -newkey rsa:4096 -nodes -sha256 -keyout certs/domain.key \
      -addext "subjectAltName = DNS:myregistry.local" \
      -x509 -days 365 -out certs/domain.crt
    sudo mkdir -p /etc/docker/certs.d/myregistry.local:5000
    sudo cp certs/domain.crt /etc/docker/certs.d/myregistry.local:5000/domain.crt
    docker run -d \
      -p 5000:5000 \
      --restart=always \
      --name registry \
      -v "$(pwd)"/certs:/certs \
      -v "$(pwd)"/registry:/var/lib/registry \
      -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
      -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
      registry:2
    ```

4. Add host name mapping on the /etc/hosts file in Harvester and pxe_server machine
  ```
  192.168.2.254 myregistry.local
  ```


5. Pull the `v0.3.3` vm-dhcp-controller images 
  ```
  docker pull rancher/harvester-vm-dhcp-controller:v0.3.3
  docker pull rancher/harvester-vm-dhcp-webhook:v0.3.3
  docker pull rancher/harvester-vm-dhcp-agent:v0.3.3
  ```

6. Tag the image and push to the private registry

  ```
  docker tag rancher/harvester-vm-dhcp-webhook:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
  docker push myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
  
  docker tag rancher/harvester-vm-dhcp-webhook:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
  docker push myregistry.local:5000/rancher/harvester-vm-dhcp-webhook:v0.3.3
  
  docker tag rancher/harvester-vm-dhcp-agent:v0.3.3 myregistry.local:5000/rancher/harvester-vm-dhcp-agent:v0.3.3
  docker push myregistry.local:5000/rancher/harvester-vm-dhcp-agent:v0.3.3
  ```

7. Prepare the chart server, first we need to clone the Harvester chart repository https://github.com/harvester/charts

8. Checkout the charts repo to `harvester-vm-dhcp-controller-0.3.3 tag` 

  ```
  cd charts/
  git checkout tags/harvester-vm-dhcp-controller-0.3.3 -b harvester-vm-dhcp-controller-0.3.3
  cd harvester-vm-dhcp-controller
  mkdir serv
  helm package . -d serv
  helm repo index serv
  cd serv/
  python3 -m http.server -b 0.0.0.0
  ```

9. Now we have prepared the private registry and chart server, we need to setup the Harvester 

10. On pxe_server, copy the content in `~/certs/domain.crt`

11. Add the domain.crt content to the Harvester additional-ca settings

  ![Image](https://github.com/user-attachments/assets/d920f857-5385-4b87-897e-934d1abf5351)
  
12.  Set the private registry in the Harvester `container-registry` setting

  ```
  {
    "Mirrors": {
      "docker.io": {
        "Endpoints": [
          "https://myregistry.local:5000"
        ],
        "Rewrites": null
      }
    },
    "Configs": {
      "myregistry.local:5000": {
        "Auth": null,
        "TLS": {
          "CAFile": "",
          "CertFile": "",
          "KeyFile": "",
          "InsecureSkipVerify": true
        }
      }
    },
    "Auths": null
  }
  ```

13. Access the Harvester node and change to root 

14. Prepare the `harvester-vm-dhcp-controller.yaml` file, refer to https://docs.harvesterhci.io/v1.4/advanced/addons/managed-dhcp#install-and-enable-the-vm-dhcp-controller-add-on

    Point the repo to the python http server http://192.168.2.254:8000 

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: Addon
  metadata:
    name: harvester-vm-dhcp-controller
    namespace: harvester-system
    labels:
      addon.harvesterhci.io/experimental: "true"
  spec:
    enabled: false
    repo: http://192.168.2.254:8000
    version: 0.3.3
    chart: harvester-vm-dhcp-controller
    valuesContent: "
  ```
  

15. Create the vm-dhcp-controller addon CRD 

  ```
  kubectl create -f harvester-vm-dhcp-controller.yaml
  ```

16. Enable the `harvester-vm-dhcp-controller` on the Harvester `Add-ons` page

**Comment by TachunLin** [comment]
During testing in the air-gapped environment built with [vagrant-pxe-airgap-harvester](https://github.com/harvester/ipxe-examples/tree/main/vagrant-pxe-airgap-harvester).

We encounter an issue that failed to upload image from file or internal accessible URL. 

![Image](https://github.com/user-attachments/assets/b3825e64-2644-4486-a542-e915f7a29d58)

Which would block the test steps to proceed.
We have create a new issue to track https://github.com/harvester/harvester/issues/7540

Thus move to ready for testing since blocked by other issue.

**Comment by starbops** [comment]
> [@starbops](https://github.com/starbops) w/ the vm-dhcp-controller is there a list of images, like an image.txt that would need to be pulled into a private docker registry prior? Alongside, hosting the chart?

@irishgordo Since harvester-vm-dhcp-controller is still in the experimental phase, the container images and the chart are not packed in the released ISO image. As a record, I'll list them down below:

- container images
  - `rancher/harvester-vm-dhcp-controller`
  - `rancher/harvester-vm-dhcp-agent`
  - `rancher/harvester-vm-dhcp-webhook`
- chart
  - [`harvester-vm-dhcp-controller`](https://github.com/harvester/charts/tree/release/charts/harvester-vm-dhcp-controller)

For air-gapped setups, you need to prepare the above container images in a private container registry and a chart server.

**Comment by irishgordo** [comment]
As another reference point, any VM/Vagrant that has the potential to be AirGapped can just leverage Hauler to condense the setup steps some.

1. as root, `curl -sfL https://get.hauler.dev | bash`
2. as standard user, build a file called `hauler-content.yaml`
```yaml
apiVersion: content.hauler.cattle.io/v1
kind: Images
metadata:
  name: "vmdhcp-controller-images"
  annotations:
    hauler.dev/registry: "docker.io"
spec:
  images:
    - name: rancher/harvester-vm-dhcp-controller:v0.3.3
    - name: rancher/harvester-vm-dhcp-webhook:v0.3.3
    - name: rancher/harvester-vm-dhcp-agent:v0.3.3
---
apiVersion: content.hauler.cattle.io/v1
kind: Images
metadata:
  name: "extra-images"
  annotations:
    hauler.dev/registry: "docker.io"
spec:
  images:
    - name: nginx:latest
---
apiVersion: content.hauler.cattle.io/v1 
kind: Charts
metadata:
  name: "vmdhcp-controller-charts"
spec:
  charts:
    - name: harvester-vm-dhcp-controller
      repoURL: https://charts.harvesterhci.io/
      version: 0.3.3
--- 
apiVersion: content.hauler.cattle.io/v1
kind: Files
metadata:
  name: additional-needed-files-prior-airgap
spec:
  files:
    - path: https://get.helm.sh/helm-v3.17.0-linux-amd64.tar.gz
    - path: https://github.com/minio/certgen/releases/latest/download/certgen-linux-amd64
---
```
3. run temporarily  `hauler store serve fileserver --port 8080` then control+c cancel it
4. airgap the VM
5. move the `/home/ubuntu/fileserver/certgen-linux-amd64` , `sudo mv -v /home/ubuntu/fileserver/certgen-linux-amd64 /usr/local/bin/certgen`
6. `sudo chmod +x /usr/local/bin/certgen` 
7.  build a `/home/ubuntu/hauler_config` that looks like this:
```yaml
version: 0.1
http:
  addr: $REPLACE-WITH-YOUR-AIRGAP-IPV4:5000
  tls:
    certificate: /home/ubuntu/public.crt
    key: /home/ubuntu/private.key
storage:
  filesystem:
    rootdirectory: /home/ubuntu/registry
    maxthreads: 100
```
8. Run certgen, `certgen -host "127.0.0.1,localhost,$IPV4,harvester-ui-airgap-registry.$IPV4.sslip.io"` ( feel free to change as needed )
9. `tmux new session -s fileserver` -> run `hauler store serve fileserver --port 8080`
10. `tmux new-session -s registry` -> run `hauler store serve registry --config /home/ubuntu/hauler_config.yaml`
11.  `cd /home/ubuntu/fileserver && tar xvf harvester-vm-dhcp-controller-0.3.3.tgz`
12. `tar xvf helm-v3.17.0-linux-amd64.tar.gz && sudo mv -v /home/ubuntu/fileserver/linux-amd64/helm /usr/local/bin/helm && sudo chmod +x /usr/local/bin/helm`
13. `cd /home/ubuntu/fileserver/harvester-vm-dhcp-controller && helm package . -d serv && helm repo index serv`
14. again modify containerd registry in Harvester:
```
{
  "Mirrors": {
    "docker.io": {
      "Endpoints": [
        "https://airgapped-vm-dhcp-registry.$IPV4.sslip.io:5000"
      ],
      "Rewrites": null
    }
  },
  "Configs": {
    "airgapped-vm-dhcp-registry.$IPV4.sslip.io:5000": {
      "Auth": null,
      "TLS": {
        "CAFile": "",
        "CertFile": "",
        "KeyFile": "",
        "InsecureSkipVerify": true
      }
    }
  },
  "Auths": null
}
```
15. Build the addon with something like:
```
apiVersion: harvesterhci.io/v1beta1
kind: Addon
metadata:
  name: harvester-vm-dhcp-controller
  namespace: harvester-system
  labels:
    addon.harvesterhci.io/experimental: "true"
spec:
  enabled: false
  repo: http://$IPV4:8080/harvester-vm-dhcp-controller/serv
  version: 0.3.3
  chart: harvester-vm-dhcp-controller
  valuesContent: ""
```

**Comment by TachunLin** [comment]
By applying the workaround to manual install LH image, we can proceed to verify this issue
https://github.com/harvester/harvester/issues/7540#issuecomment-2652598991

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.5-7e12eb95-head` (25/02/11). Close this issue.

### Pre-requisite setup

* Alternative 1: Configure with private registry and simple http chart server 
  - https://github.com/harvester/harvester/issues/6942#issuecomment-2643065019
* Alternative 2: Configure with hauler 
  - https://github.com/harvester/harvester/issues/6942#issuecomment-2649514067

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Can install and enable vm-dhcp-controller in air-gapped environment $~~$</summary>

  ![Image](https://github.com/user-attachments/assets/6cba4ce9-3a40-47d8-9a2a-54e5bf8886ee)
  
 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Can create the ippool network object and condition is good $~~$</summary>
 
1. Check ippools network object condition is good


    ```
    david-10-115-254-106:~ # kubectl get ippools.network
    NAME       NETWORK            AVAILABLE   USED   REGISTERED   CACHEREADY   AGENTREADY   AGE
    net-2011   default/net-2011   8           0      True         True         True         6m6s
    ```
    
    ![Image](https://github.com/user-attachments/assets/6ec90576-3683-43bf-8b56-7ce8c92f0108)

2. dhcpserver-agent in `Running` state
    ![Image](https://github.com/user-attachments/assets/18f8a2e9-9fc3-44c7-afc8-e4d1acf7aca8)
    
 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ VM can get the IP address of the vm-dhcp-controller ippool object $~~$</summary>
 
3. Check the IP address is correctly configured on the network interface of the virtual machine

  ![Image](https://github.com/user-attachments/assets/fc98d028-6608-46f3-9ea4-eef2020eca9a)
  
  ![Image](https://github.com/user-attachments/assets/85f6fec8-d6b2-4916-9da6-9c6313a5b243)
  
 </details>

### Test Information
* Test Environment: Single node harvester on baremetal and kvm machine
* Harvester version: v1.1-340486a3-head (23/11/20)

### Verify Steps
<details><summary> Create IPPool object on the virtual machine network and set virtual machine network config on it </summary>

Refer to: 
https://docs.harvesterhci.io/v1.4/advanced/addons/managed-dhcp/#install-and-enable-the-vm-dhcp-controller-add-on

1. Install the harvester-vm-dhcp-controller experimental add-on
2. Enable the add-on
3. Create a cluster network
4. Create a network config with at least one extra nic
5. Create a virtual machine network with available vlan id

  ![Image](https://github.com/user-attachments/assets/35092048-144c-48db-ae53-2f107d279a93)
  
6. Create an IPPool object

  ```
  cat <<EOF | kubectl apply -f -
  apiVersion: network.harvesterhci.io/v1alpha1
  kind: IPPool
  metadata:
    name: net-2011
    namespace: default
  spec:
    ipv4Config:
      serverIP: 192.168.201.77
      cidr: 192.168.201.0/24
      pool:
        start: 192.168.201.81
        end: 192.168.201.90
        exclude:
        - 192.168.201.81
        - 192.168.201.90
      router: 192.168.201.1
      dns:
      - 1.1.1.1
      leaseTime: 300
    networkName: default/net-2011
  EOF
  ```


7. Ensure the IPPool object's conditions are all good, especially the AgentReady one 

  ```
  kubectl get ippool.network.harvesterhci.io -A
  ```

8. Check the `default-net-2011-agent` pod status in K9s
    
9. Create a VM
10. Set the virtual machine network to `net-2011`
11. Ensure the VM gets the IP address recorded in the corresponding VirtualMachineNetworkConfig object


</details>
